File: contracts/stubs/adapter-pack.ts
/**
 * Adapter-pack contract, TypeScript form — contracts-v1.
 *
 * See ../adapter-pack.md for semantics. Real packs live in-tree in keel-node;
 * this interface is the frozen shape they implement.
 */

export interface Detection {
  matched: boolean;
  /** e.g. "undici", "openai", "ai" (Vercel AI SDK) */
  name?: string;
  /** installed version, undefined if unknown */
  version?: string;
  confidence?: "pinned" | "best_effort";
}

export interface Seam {
  /** e.g. "undici.Dispatcher#dispatch" */
  patchPoint: string;
  /** the documented upstream API this relies on */
  upstreamApi: string;
  /** printed verbatim by `keel doctor` */
  whyStable: string;
}

export interface TargetDecl {
  /** target id or pattern, e.g. "llm:openai", "mcp:<server>" */
  pattern: string;
  kind: "host" | "function" | "llm" | "tool" | "mcp";
  /** how `idempotent` is derived at the seam */
  idempotencyRule: string;
  /** how `args_hash` is derived at the seam */
  argsHashRule: string;
}

/**
 * The four operations every pack implements. No retry/backoff/breaker logic
 * lives here — all behavior flows through the core.
 */
export interface AdapterPack {
  detect(): Detection;
  seams(): Seam[];
  targets(): TargetDecl[];
  /**
   * Policy fragment (keel.toml JSON form, per policy.schema.json), merged
   * UNDER user configuration.
   */
  defaults(): Record<string, unknown>;
}

File: contracts/stubs/adapter_pack.py
"""Adapter-pack contract, Python form — contracts-v1.

See ../adapter-pack.md for semantics. Real packs live in-tree in keel-py;
this Protocol is the frozen shape they implement.
"""

from dataclasses import dataclass, field
from typing import Any, Literal, Protocol


@dataclass(frozen=True)
class Detection:
    matched: bool
    name: str = ""                      # e.g. "httpx", "openai", "google-adk"
    version: str = ""                   # installed version, "" if unknown
    confidence: Literal["pinned", "best_effort"] = "best_effort"


@dataclass(frozen=True)
class Seam:
    patch_point: str                    # e.g. "httpx.HTTPTransport.handle_request"
    upstream_api: str                   # the documented API this relies on
    why_stable: str                     # printed verbatim by `keel doctor`


@dataclass(frozen=True)
class TargetDecl:
    pattern: str                        # target id or pattern, e.g. "llm:openai"
    kind: Literal["host", "function", "llm", "tool", "mcp"]
    idempotency_rule: str               # how `idempotent` is derived at the seam
    args_hash_rule: str                 # how `args_hash` is derived at the seam


class AdapterPack(Protocol):
    """The four operations every pack implements. No retry/backoff/breaker
    logic lives here — all behavior flows through the core."""

    def detect(self) -> Detection: ...

    def seams(self) -> list[Seam]: ...

    def targets(self) -> list[TargetDecl]: ...

    def defaults(self) -> dict[str, Any]:
        """Policy fragment (keel.toml JSON form, per policy.schema.json),
        merged UNDER user configuration."""
        ...


__all__ = ["AdapterPack", "Detection", "Seam", "TargetDecl"]
